Check-in command support

kennylevinsen opened the pull request:
This implements support for receiving and responding to check-in commands. If there are any pending messages, fast poll is requested and the queue is emptied. Otherwise, fast poll is declined.

Notes:
- The device configure is still responsible for binding the cluster, as the messages will otherwise never arrive.
- Implicit check-in is disabled if we receive a check-in message, but maybe it would be nicer if it was controlled together with the binding setup (device.enablePollClusters())?
- The check-in command still bubbles up as a message, so if the device lacks a genPollCtrl commandCheckIn converter, zigbee2mqtt will log an error. We could let the message be consumed, but the "checkin_presence" converter currently relies on it.

Comments welcome.

Related to https://github.com/Koenkk/zigbee-herdsman/issues/445

Koenkk commented:
> Implicit check-in is disabled if we receive a check-in message, but maybe it would be nicer if it was controlled together with the binding setup (device.enablePollClusters())?

Good point, we can do this in the `configure()` and save this in the database, now `useImplicitCheckin` is not persisted across restarts.

> The check-in command still bubbles up as a message, so if the device lacks a genPollCtrl commandCheckIn converter, zigbee2mqtt will log an error. We could let the message be consumed, but the "checkin_presence" converter currently relies on it.

Can be solved easily by adding `genPollCtrl` to https://github.com/Koenkk/zigbee2mqtt/blob/b3e9afaf3ee55e2d62f6775dc0c62e8550b8fef1/lib/extension/receive.ts#L117


I see the tests are currently failing, can you update these and make sure the new code is covered? Use `npm tst` and `npm run test-with-coverage`.

kennylevinsen commented:
> Good point, we can do this in the configure() and save this in the database, now useImplicitCheckin is not persisted across restarts.

Take a look at `configurePollControl` and see what you think. I find the interface slightly awkward, but on the other hand I feel like it would be very manual to use poll control if people had to bind and flip implicitCheckin themselves.

> I see the tests are currently failing, can you update these and make sure the new code is covered? Use npm tst and npm run test-with-coverage.

Done, although tests fail on master right now.

Koenkk commented:
Sorry for the failing tests, I've fixed this now in master.
Reviewed the code, looks good to me so if you rebase and CI checks pass I will merge this.

kennylevinsen commented:
> Sorry for the failing tests, I've fixed this now in master.

It happens - I was going to submit a fix, but you beat me to it. :)

Rebased. Now we're just failing a lint on a file that hasn't been changed.

Koenkk commented:
> Rebased. Now we're just failing a lint on a file that hasn't been changed.

It fails in `src/zcl/zclFrame.ts` which has been changed

kennylevinsen commented:
> It fails in `src/zcl/zclFrame.ts` which has been changed

... Apologies, I must have forgotten to put my eyeballs back in. You're right, and fixed.

Koenkk commented:
Haha 👀 , thanks!
